From 825b0432ff61a21c0dc19f7a66e848822f6da6a7 Mon Sep 17 00:00:00 2001
From: ctang <ctang@cloudera.com>
Date: Sat, 9 Jul 2016 22:36:56 -0400
Subject: [PATCH 0805/1164] HIVE-14173: NPE was thrown after enabling
 directsql in the middle of session (Chaoyu Tang,
 reviewed by Sergey Shelukhin)

Change-Id: I42090797f767716e8e242fb493f22b42aa60ec27
---
 .../apache/hadoop/hive/metastore/ObjectStore.java  |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java b/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
index 7a6a1d6..78b4c6a 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java
@@ -2594,6 +2594,10 @@ public GetHelper(String dbName, String tblName, boolean allowSql, boolean allowJ
       // the fallback from failed SQL to JDO is not possible.
       boolean isConfigEnabled = HiveConf.getBoolVar(getConf(), ConfVars.METASTORE_TRY_DIRECT_SQL)
           && (HiveConf.getBoolVar(getConf(), ConfVars.METASTORE_TRY_DIRECT_SQL_DDL) || !isInTxn);
+      if (isConfigEnabled && directSql == null) {
+        directSql = new MetaStoreDirectSql(pm, getConf());
+      }
+
       if (!allowJdo && isConfigEnabled && !directSql.isCompatibleDatastore()) {
         throw new MetaException("SQL is not operational"); // test path; SQL is enabled and broken.
       }
-- 
1.7.9.5

